rfd/0137-unified-resource-view.md

```diff
@@ -0,0 +1,121 @@
+
+The one biggish change is, currently, `collector.resourceKind()` returns a single type as a string and when the watcher runs `resourceWatcher.watch()` it creates a slice containing the single Kind returned. We can make `resourceKind()` take ownership of supplying the slice instead of `watch`, which would allow us to return the slice of all desired kinds to watch
+```go
+func (u *unifiedResourceCollector) resourceKind() []types.WatchKind {
+	return []types.WatchKind{
+		{Kind: types.KindNode},
+		{Kind: types.KindDatabaseServer},
+		{Kind: types.KindAppServer},
+		{Kind: types.KindWindowsDesktop},
+	}
+}
+```
+The above example can be expanded to include any Kinds we may need/want to add in our Unified Resource view.
+
+When the watcher inits, we populate the btree with all the available items on startup
+```go
+func (u *uiResourceCollector) getResourcesAndUpdateCurrent(ctx context.Context) error {
+	err := u.getAndUpdateNodes(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = u.getAndUpdateDatabases(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	//...
+
+	u.defineCollectorAsInitialized()
+	return nil
+}
+```
+
+If the collector is uninitialized or stale, we will resort to the upstream and send Nodes only.
+
+### ListUnifiedResources grpc endppint
+We will add a new grpc endpoint, `ListUnifiedResources`. It will function pretty closes to how `ListResources` works, except without needing to differentiate between `RequestType`, as every request is a unified resource request. Originally, the thought of just extending `ListResources`, but because of the many new filters that will be added to the request, we don't want to muddy up the `ListResourcesRequest`. Some of the new filters include which types to include (can be multiple), subtypes, etc etc. 
+
+This endpoint will pull resources from our unified resource cache and run through the same filtering/rbac as `ListResources`
+
+### Backward Compatibility
+
+We will leave `ListResources` alone and keep the old individual resource endpoints available until we've vetted the unified resource view. We will remove the individual navigation items from the web ui, but leave the routes/api endpoints until we've gained confidence in the new flow. 
+
+### UX
+The UI for the web will be changing drastically given the fact that all resources will be on the same page and we will be using a Card layout rather than the current Table. However, besides the visual changes, the actual resources listed will still be interacted with the same way. You can still click "Connect" on servers and databases, or Login for apps. The individual functionality for each "Item" will remain the same. Labels will be truncated but viewable with an expand button.
```

Comment:
just curious how rbac plays into this. what if a user didn't have access to for example nodes? how will we let the user know that they don't have access? (do we want to?)

i think previously, if a user didn't have "list" verb, the tab won't render. I think we are trying to move away from "silently" not rendering things and be more loud about it (eg: you don't have access to nodes, reach out to your teleport admin to request additional perm etc) but i'm not sure

Reply:
Currently, we just show what is available to user. I believe it's a "silent" fail if they don't have access to something like Kubes. cc: @roraback

Reply:
Correct. Users could still discover types through the filter options, and if they filter or search for a type, such as kubernetes, the empty state should give them an indication. The current blank resource screen doesn't seem particularly insightful in the current state, anyway: it just tells them that _either_ there is nothing to see _or_ they don't have permissions, which they could probably figure out for themselves if they know anything about their company's infrastructure. Please LMK if I'm missing something, though!

This does remind me that design for the empty state was one of the casualties of time constraints—we should be careful about our messaging to ensure it doesn't give false impressions.